starky/src/compressor12/compressor12_pil.rs

```diff
@@ -49,51 +101,340 @@ namespace Compressor(%N);
+    POSEIDON12 * (a[10]' - (41*a0_R + 16*a1_R +  2*a2_R + 28*a3_R + 13*a4_R + 13*a5_R + 39*a6_R + 18*a7_R + 34*a8_R + 20*a9_R + 17*a10_R + 15*a11_R)) = 0;
+    POSEIDON12 * (a[11]' - (15*a0_R + 41*a1_R + 16*a2_R +  2*a3_R + 28*a4_R + 13*a5_R + 13*a6_R + 39*a7_R + 18*a8_R + 34*a9_R + 20*a10_R + 17*a11_R)) = 0;
+
+
+    // CMULADD GATE
+    pol ca0 = (a[0] + C[0])*C[9];
+    pol ca1 = (a[1] + C[1])*C[9];
+    pol ca2 = (a[2] + C[2])*C[9];
+    pol ca3 = a[3] + C[3];
+    pol ca4 = a[4] + C[4];
+    pol ca5 = a[5] + C[5];
+    pol ca6 = (a[6] + C[6])*C[10];
+    pol ca7 = (a[7] + C[7])*C[10];
+    pol ca8 = (a[8] + C[8])*C[10];
+    pol ca9 = a[9];
+    pol ca10 = a[10];
+    pol ca11 = a[11];
+
+    pol cA = (ca0 + ca1)  * (ca3 + ca4);
+    pol cB = (ca0 + ca2)  * (ca3 + ca5);
+    pol cC = (ca1 + ca2)  * (ca4 + ca5);
+    pol cD = ca0*ca3;
+    pol cE = ca1*ca4;
+    pol cF = ca2*ca5;
+
+    CMULADD * (ca9 - (cC + cD - cE - cF) - ca6) = 0;
+    CMULADD * (ca10 - (cA + cC - 2*cE - cD) - ca7) = 0;
+    CMULADD * (ca11 - (cB - cD + cE) - ca8) = 0;
+
+    // FFT4
+    
+    pol g0 = C[0]*a[0] + C[1]*a[3] + C[2]*a[6] + C[3]*a[9]  + C[6]*a[0] + C[7]*a[3];
+    pol g1 = C[0]*a[1] + C[1]*a[4] + C[2]*a[7] + C[3]*a[10] + C[6]*a[1] + C[7]*a[4];
+    pol g2 = C[0]*a[2] + C[1]*a[5] + C[2]*a[8] + C[3]*a[11] + C[6]*a[2] + C[7]*a[5];
+
+    pol g3 = C[0]*a[0] - C[1]*a[3] + C[4]*a[6] - C[5]*a[9]  + C[6]*a[0] - C[7]*a[3];
+    pol g4 = C[0]*a[1] - C[1]*a[4] + C[4]*a[7] - C[5]*a[10] + C[6]*a[1] - C[7]*a[4];
+    pol g5 = C[0]*a[2] - C[1]*a[5] + C[4]*a[8] - C[5]*a[11] + C[6]*a[2] - C[7]*a[5];
+
+    pol g6 = C[0]*a[0] + C[1]*a[3] - C[2]*a[6] - C[3]*a[9]  + C[6]*a[6] + C[8]*a[9];
+    pol g7 = C[0]*a[1] + C[1]*a[4] - C[2]*a[7] - C[3]*a[10] + C[6]*a[7] + C[8]*a[10];
+    pol g8 = C[0]*a[2] + C[1]*a[5] - C[2]*a[8] - C[3]*a[11] + C[6]*a[8] + C[8]*a[11];
+
+    pol g9  = C[0]*a[0] - C[1]*a[3] - C[4]*a[6] + C[5]*a[9]  + C[6]*a[6] - C[8]*a[9];
+    pol g10 = C[0]*a[1] - C[1]*a[4] - C[4]*a[7] + C[5]*a[10] + C[6]*a[7] - C[8]*a[10];
+    pol g11 = C[0]*a[2] - C[1]*a[5] - C[4]*a[8] + C[5]*a[11] + C[6]*a[8] - C[8]*a[11];
+
+    FFT4 * (a[0]' - g0) = 0;
+    FFT4 * (a[1]' - g1) = 0;
+    FFT4 * (a[2]' - g2) = 0;
+    FFT4 * (a[3]' - g3) = 0;
+    FFT4 * (a[4]' - g4) = 0;
+    FFT4 * (a[5]' - g5) = 0;
+    FFT4 * (a[6]' - g6) = 0;
+    FFT4 * (a[7]' - g7) = 0;
+    FFT4 * (a[8]' - g8) = 0;
+    FFT4 * (a[9]' - g9) = 0;
+    FFT4 * (a[10]' - g10) = 0;
+    FFT4 * (a[11]' - g11) = 0;
+
+       "#
+    );
+
+    // CMulAdd
+    let mut c_mul_add = |r0: &str,
+                         r1: &str,
+                         r2: &str,
+                         a0: &str,
+                         a1: &str,
+                         a2: &str,
+                         b0: &str,
+                         b1: &str,
+                         b2: &str,
+                         c0: &str,
+                         c1: &str,
+                         c2: &str| {
+        res.push_str(
```

Comment:
Can we merge the line 259, 269, 285, and more code lines into one string, and then push the string into res?

Reply:
Maybe keeping this as a function is better for maintenance.